File: web/lib/api/agent-lens.ts
import { apiClient } from './http';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LensStatCard {
  total_sessions: number;
  total_requests: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  avg_latency_ms: number;
  total_cost_usd: number;
  avg_cost_per_session: number;
  failure_rate: number;
  total_tool_calls: number;
  failed_tool_calls: number;
  avg_tool_duration_ms: number;
}

export interface LensOverviewResponse {
  stats: LensStatCard;
  period_start: string;
  period_end: string;
}

export interface LensSessionRow {
  id: string;
  name: string;
  status: string;
  model_name: string | null;
  total_tokens: number;
  cost_usd: number;
  message_count: number;
  created_at: string | null;
}

export interface LensSessionsResponse {
  sessions: LensSessionRow[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface LensTimelineEvent {
  event_type: 'user_message' | 'llm_call' | 'tool_call' | 'assistant_message' | 'tool_pruning';
  id: string;
  timestamp: string | null;
  sequence?: number;
  // user_message / assistant_message fields
  role?: string;
  content?: string;
  content_preview?: string;
  token_count?: number;
  input_tokens?: number;
  output_tokens?: number;
  latency_ms?: number;
  total_latency_ms?: number;
  total_cost_usd?: number;
  status?: string;
  error?: string;
  // llm_call fields
  model?: string;
  call_index?: number;
  cost_usd?: number;
  response_preview?: string;
  // llm_call input inspection fields
  system_prompt_preview?: string;
  messages_json?: string;  // JSON string: list of chat messages sent to LLM
  tools_json?: string;     // JSON string: list of tool definitions sent to LLM
  // llm_call prompt cache fields
  cache_read_tokens?: number;
  cache_creation_tokens?: number;
  // llm_call context window pressure (0–100 %)
  context_utilization_pct?: number;
  // tool call fields
  tool_name?: string;
  tool_args?: string;  // JSON string stored in ES text field
  tool_result?: string;  // JSON string stored in ES text field
  success?: boolean;
  duration_ms?: number;
  retry_count?: number;
  error_message?: string;
  // tool_pruning fields
  turn_index?: number;
  tool_results_count?: number;
  tool_results_pruned?: number;
  data_bearing_pruned?: number;
  chars_saved?: number;
  estimated_tokens_saved?: number;
  original_chars?: number;
  pruned_chars?: number;
}

export interface LensSessionDetailResponse {
  conversation_id: string;
  name: string;
  status: string;
  created_at: string | null;
  timeline: LensTimelineEvent[];
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  total_cost_usd: number;
  total_tool_calls: number;
  failed_tool_calls: number;
  message_count: number;
}

export interface LensModelBreakdown {
  model_name: string | null;
  tokens: number;
  cost_usd: number;
  calls: number;
}

export interface LensTokenDistributionResponse {
  by_model: LensModelBreakdown[];
  prompt_vs_completion: {
    input_tokens: number;
    output_tokens: number;
  };
}

export interface AlertResponse {
  id: string;
  agent_id: string;
  name: string;
  metric: string;
  condition: string;
  threshold: number;
  window_minutes: number;
  notify_method: string;
  notify_config: Record<string, unknown> | null;
  is_active: boolean;
  last_triggered_at: string | null;
  created_at: string | null;
}

export interface LensToolStat {
  tool_name: string;
  total_calls: number;
  success_count: number;
  failure_count: number;
  success_rate: number;  // 0.0 – 1.0
  avg_duration_ms: number;
  max_duration_ms: number;
  total_retries: number;
}

export interface LensToolAnalyticsResponse {
  tools: LensToolStat[];
  total_calls: number;
  total_failures: number;
  overall_success_rate: number;
}

export interface LensToolROIStat {
  tool_name: string;
  total_calls: number;
  calls_per_session: number;
  sessions_with: number;
  sessions_without: number;
  error_rate: number;
  avg_cost_with: number;
  avg_cost_without: number;
  cost_impact_pct: number;        // positive = adds cost
  avg_llm_calls_with: number;
  avg_llm_calls_without: number;
  completion_rate_with: number;   // 0.0–1.0
  completion_rate_without: number;
  completion_lift: number;        // positive = tool helps resolution
  verdict: 'keep' | 'review' | 'remove' | 'neutral';
}

export interface LensToolROIResponse {
  tools: LensToolROIStat[];
  total_sessions: number;
  period_start: string;
  period_end: string;
}

export interface AlertCreateBody {
  name: string;
  metric: 'failure_rate' | 'cost_per_session' | 'avg_latency_ms' | 'total_cost_usd' | 'token_count';
  condition: 'gt' | 'lt';
  threshold: number;
  window_minutes?: number;
  notify_method?: 'email' | 'webhook';
  notify_config?: Record<string, unknown>;
  is_active?: boolean;
}

export type AlertUpdateBody = Partial<AlertCreateBody>;

export type LensRange = '24h' | '7d' | '30d' | '90d';

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function getLensOverview(agentSlug: string, range: LensRange = '7d'): Promise<LensOverviewResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/overview?range=${range}`);
}

export async function getLensTokenDistribution(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensTokenDistributionResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/token-distribution?range=${range}`);
}

export async function getLensSessions(
  agentSlug: string,
  range: LensRange = '7d',
  page = 1,
  pageSize = 20
): Promise<LensSessionsResponse> {
  return apiClient.request(
    'GET',
    `/api/v1/agents/${agentSlug}/lens/sessions?range=${range}&page=${page}&page_size=${pageSize}`
  );
}

export async function getLensSessionDetail(
  agentSlug: string,
  sessionId: string
): Promise<LensSessionDetailResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/sessions/${sessionId}`);
}

export async function getLensToolAnalytics(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensToolAnalyticsResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/tool-analytics?range=${range}`);
}

export async function getLensAlerts(agentSlug: string): Promise<AlertResponse[]> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/alerts`);
}

export async function createLensAlert(agentSlug: string, body: AlertCreateBody): Promise<AlertResponse> {
  return apiClient.request('POST', `/api/v1/agents/${agentSlug}/lens/alerts`, body);
}

export async function updateLensAlert(
  agentSlug: string,
  alertId: string,
  body: AlertUpdateBody
): Promise<AlertResponse> {
  return apiClient.request('PUT', `/api/v1/agents/${agentSlug}/lens/alerts/${alertId}`, body);
}

export async function deleteLensAlert(agentSlug: string, alertId: string): Promise<void> {
  return apiClient.request('DELETE', `/api/v1/agents/${agentSlug}/lens/alerts/${alertId}`);
}

export async function getLensToolROI(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensToolROIResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/tool-roi?range=${range}`);
}

// ---------------------------------------------------------------------------
// Compaction analytics
// ---------------------------------------------------------------------------

export interface LensPruningTrendPoint {
  date: string;
  pruning_events: number;
  tokens_saved: number;
  data_bearing_pruned: number;
}

export interface LensCompactionAnalyticsResponse {
  total_sessions: number;
  sessions_with_pruning: number;
  pruning_rate: number;
  pruning_events_count: number;
  total_tokens_saved: number;
  avg_tokens_saved_per_session: number;
  data_bearing_pruned_total: number;
  total_results_pruned: number;
  total_results_evaluated: number;
  avg_pruning_ratio: number;
  trend: LensPruningTrendPoint[];
  period_start: string;
  period_end: string;
}

export async function getLensCompaction(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensCompactionAnalyticsResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/compaction?range=${range}`);
}

// ---------------------------------------------------------------------------
// Cache analytics
// ---------------------------------------------------------------------------

export interface LensCacheTrendPoint {
  date: string;
  input_tokens: number;
  cache_read_tokens: number;
  cache_creation_tokens: number;
  cache_hit_rate: number;
}

export interface LensCacheAnalyticsResponse {
  total_llm_calls: number;
  total_input_tokens: number;
  total_cache_read_tokens: number;
  total_cache_creation_tokens: number;
  cache_hit_rate: number;
  estimated_savings_usd: number;
  avg_context_utilization_pct: number;
  max_context_utilization_pct: number;
  high_pressure_calls: number;
  trend: LensCacheTrendPoint[];
  period_start: string;
  period_end: string;
}

export async function getLensCache(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensCacheAnalyticsResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/cache?range=${range}`);
}

// ---------------------------------------------------------------------------
// Quality (satisfaction + outcomes)
// ---------------------------------------------------------------------------

export interface LensQualityChannelBreakdown {
  channel: string;
  satisfaction_rate: number;
  rated_count: number;
}

export interface LensSatisfactionStats {
  total_rated: number;
  thumbs_up: number;
  thumbs_down: number;
  satisfaction_rate: number;
  by_channel: LensQualityChannelBreakdown[];
}

export interface LensOutcomeStats {
  total_sessions: number;
  success: number;
  partial: number;
  failure: number;
  unknown: number;
  success_rate: number;
  explicit_rate: number;
  implicit_rate: number;
}

export interface LensDislikedTool {
  tool_name: string;
  dislike_rate: number;
  dislike_count: number;
}

export interface LensQualityResponse {
  period_start: string;
  period_end: string;
  satisfaction: LensSatisfactionStats;
  outcomes: LensOutcomeStats;
  top_disliked_tools: LensDislikedTool[];
}

export async function getLensQuality(
  agentSlug: string,
  range: LensRange = '7d'
): Promise<LensQualityResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/quality?range=${range}`);
}

// ---------------------------------------------------------------------------
// Eval history
// ---------------------------------------------------------------------------

export interface LensEvalRunRow {
  run_id: string;
  dataset_id: string;
  dataset_name: string | null;
  ran_at: string | null;
  status: string;
  total_cases: number;
  passed: number;
  failed: number;
  pass_rate: number;
  avg_score: number;
  avg_latency_ms: number;
}

export interface LensEvalHistoryResponse {
  runs: LensEvalRunRow[];
  total: number;
}

export async function getLensEvalHistory(agentSlug: string): Promise<LensEvalHistoryResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/eval-history`);
}

// ---------------------------------------------------------------------------
// Session graph
// ---------------------------------------------------------------------------

export interface LensGraphNode {
  id: string;
  event_type: string;
  label: string;
  status: 'normal' | 'warning' | 'error' | 'info';
  metadata: Record<string, unknown> | null;
}

export interface LensGraphEdge {
  source: string;
  target: string;
}

export interface LensSessionGraphResponse {
  conversation_id: string;
  nodes: LensGraphNode[];
  edges: LensGraphEdge[];
}

export async function getLensSessionGraph(
  agentSlug: string,
  sessionId: string
): Promise<LensSessionGraphResponse> {
  return apiClient.request('GET', `/api/v1/agents/${agentSlug}/lens/sessions/${sessionId}/graph`);
}

File: api/src/schemas/agent_lens.py
"""
Agent Lens Schemas

Request/response Pydantic models for the observability API.
"""

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Overview
# ---------------------------------------------------------------------------


class LensStatCard(BaseModel):
    total_sessions: int
    total_requests: int
    total_tokens: int
    input_tokens: int
    output_tokens: int
    avg_latency_ms: int
    total_cost_usd: float
    avg_cost_per_session: float
    failure_rate: float
    total_tool_calls: int
    failed_tool_calls: int
    avg_tool_duration_ms: int
    satisfaction_rate: float | None = None
    session_success_rate: float | None = None
    latest_eval_pass_rate: float | None = None


class LensOverviewResponse(BaseModel):
    stats: LensStatCard
    period_start: str
    period_end: str


# ---------------------------------------------------------------------------
# Sessions list
# ---------------------------------------------------------------------------


class LensSessionRow(BaseModel):
    id: str
    name: str
    status: str
    model_name: str | None = None
    total_tokens: int
    cost_usd: float
    message_count: int
    created_at: str | None = None


class LensSessionsResponse(BaseModel):
    sessions: list[LensSessionRow]
    total: int
    page: int
    page_size: int
    total_pages: int


# ---------------------------------------------------------------------------
# Session detail / timeline
# ---------------------------------------------------------------------------


class LensTimelineEvent(BaseModel):
    event_type: Literal["user_message", "llm_call", "tool_call", "assistant_message", "tool_pruning"]
    id: str
    timestamp: str | None = None
    sequence: int | None = None
    # user_message / assistant_message fields
    role: str | None = None
    content: str | None = None
    content_preview: str | None = None
    token_count: int | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    latency_ms: int | None = None
    total_latency_ms: int | None = None
    total_cost_usd: float | None = None
    status: str | None = None
    error: str | None = None
    # llm_call fields
    model: str | None = None
    call_index: int | None = None
    cost_usd: float | None = None
    response_preview: str | None = None
    # llm_call input inspection fields
    system_prompt_preview: str | None = None
    messages_json: str | None = None  # JSON string: list of chat messages sent to LLM
    tools_json: str | None = None  # JSON string: list of tool definitions sent to LLM
    # llm_call prompt cache fields
    cache_read_tokens: int | None = None
    cache_creation_tokens: int | None = None
    # llm_call context window pressure (0–100 %)
    context_utilization_pct: float | None = None
    # tool call fields
    tool_name: str | None = None
    tool_args: str | None = None  # JSON string stored in ES text field
    tool_result: str | None = None  # JSON string stored in ES text field
    success: bool | None = None
    duration_ms: int | None = None
    retry_count: int | None = None
    error_message: str | None = None
    # tool_pruning fields
    turn_index: int | None = None
    tool_results_count: int | None = None
    tool_results_pruned: int | None = None
    data_bearing_pruned: int | None = None  # > 0 means data-holding results were trimmed
    chars_saved: int | None = None
    estimated_tokens_saved: int | None = None
    original_chars: int | None = None
    pruned_chars: int | None = None


class LensSessionDetailResponse(BaseModel):
    conversation_id: str
    name: str
    status: str
    created_at: str | None = None
    timeline: list[LensTimelineEvent]
    total_tokens: int
    input_tokens: int
    output_tokens: int
    total_cost_usd: float
    total_tool_calls: int
    failed_tool_calls: int
    message_count: int


# ---------------------------------------------------------------------------
# Token distribution
# ---------------------------------------------------------------------------


class LensModelBreakdown(BaseModel):
    model_name: str | None
    tokens: int
    cost_usd: float
    calls: int


class LensPromptVsCompletion(BaseModel):
    input_tokens: int
    output_tokens: int


class LensTokenDistributionResponse(BaseModel):
    by_model: list[LensModelBreakdown]
    prompt_vs_completion: LensPromptVsCompletion


# ---------------------------------------------------------------------------
# Tool analytics
# ---------------------------------------------------------------------------


class LensToolStat(BaseModel):
    tool_name: str
    total_calls: int
    success_count: int
    failure_count: int
    success_rate: float  # 0.0 – 1.0
    avg_duration_ms: int
    max_duration_ms: int
    total_retries: int


class LensToolAnalyticsResponse(BaseModel):
    tools: list[LensToolStat]
    total_calls: int
    total_failures: int
    overall_success_rate: float


# ---------------------------------------------------------------------------
# Tool ROI
# ---------------------------------------------------------------------------


class LensToolROIStat(BaseModel):
    tool_name: str
    total_calls: int
    calls_per_session: float
    sessions_with: int
    sessions_without: int
    error_rate: float
    avg_cost_with: float
    avg_cost_without: float
    cost_impact_pct: float  # positive = adds cost, negative = saves cost
    avg_llm_calls_with: float
    avg_llm_calls_without: float
    completion_rate_with: float  # 0.0–1.0
    completion_rate_without: float  # 0.0–1.0
    completion_lift: float  # positive = tool helps resolution
    verdict: str  # keep | review | remove | neutral


class LensToolROIResponse(BaseModel):
    tools: list[LensToolROIStat]
    total_sessions: int
    period_start: str
    period_end: str


# ---------------------------------------------------------------------------
# Alerts
# ---------------------------------------------------------------------------


class LensQualityChannelBreakdown(BaseModel):
    channel: str
    satisfaction_rate: float
    rated_count: int


class LensSatisfactionStats(BaseModel):
    total_rated: int
    thumbs_up: int
    thumbs_down: int
    satisfaction_rate: float
    by_channel: list[LensQualityChannelBreakdown]


class LensOutcomeStats(BaseModel):
    total_sessions: int
    success: int
    partial: int
    failure: int
    unknown: int
    success_rate: float
    explicit_rate: float
    implicit_rate: float


class LensDislikedTool(BaseModel):
    tool_name: str
    dislike_rate: float
    dislike_count: int


class LensQualityResponse(BaseModel):
    period_start: str
    period_end: str
    satisfaction: LensSatisfactionStats
    outcomes: LensOutcomeStats
    top_disliked_tools: list[LensDislikedTool]


class LensEvalRunRow(BaseModel):
    run_id: str
    dataset_id: str
    dataset_name: str | None = None
    ran_at: str | None = None
    status: str
    total_cases: int
    passed: int
    failed: int
    pass_rate: float
    avg_score: float
    avg_latency_ms: int


class LensEvalHistoryResponse(BaseModel):
    runs: list[LensEvalRunRow]
    total: int


class AlertCreateBody(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    metric: Literal[
        "failure_rate",
        "cost_per_session",
        "avg_latency_ms",
        "total_cost_usd",
        "token_count",
        "satisfaction_rate",
        "session_success_rate",
    ]
    condition: Literal["gt", "lt"]
    threshold: float
    window_minutes: int = Field(default=60, ge=1, le=10080)
    notify_method: Literal["email", "webhook"] = "email"
    notify_config: dict[str, Any] | None = None
    is_active: bool = True


class AlertUpdateBody(BaseModel):
    name: str | None = None
    metric: (
        Literal[
            "failure_rate",
            "cost_per_session",
            "avg_latency_ms",
            "total_cost_usd",
            "token_count",
            "satisfaction_rate",
            "session_success_rate",
        ]
        | None
    ) = None
    condition: Literal["gt", "lt"] | None = None
    threshold: float | None = None
    window_minutes: int | None = None
    notify_method: Literal["email", "webhook"] | None = None
    notify_config: dict[str, Any] | None = None
    is_active: bool | None = None


class AlertResponse(BaseModel):
    id: str
    agent_id: str
    name: str
    metric: str
    condition: str
    threshold: float
    window_minutes: int
    notify_method: str
    notify_config: dict[str, Any] | None = None
    is_active: bool
    last_triggered_at: str | None = None
    created_at: str | None = None


# ---------------------------------------------------------------------------
# Compaction / pruning analytics
# ---------------------------------------------------------------------------


class LensPruningTrendPoint(BaseModel):
    date: str
    pruning_events: int
    tokens_saved: int
    data_bearing_pruned: int


class LensCompactionAnalyticsResponse(BaseModel):
    total_sessions: int
    sessions_with_pruning: int
    pruning_rate: float  # 0.0–1.0
    pruning_events_count: int
    total_tokens_saved: int
    avg_tokens_saved_per_session: float
    data_bearing_pruned_total: int  # how many data-bearing results were trimmed
    total_results_pruned: int
    total_results_evaluated: int
    avg_pruning_ratio: float  # pruned/evaluated ratio
    trend: list[LensPruningTrendPoint]
    period_start: str
    period_end: str


# ---------------------------------------------------------------------------
# Cache analytics
# ---------------------------------------------------------------------------


class LensCacheTrendPoint(BaseModel):
    date: str
    input_tokens: int
    cache_read_tokens: int
    cache_creation_tokens: int
    cache_hit_rate: float


class LensCacheAnalyticsResponse(BaseModel):
    total_llm_calls: int
    total_input_tokens: int
    total_cache_read_tokens: int
    total_cache_creation_tokens: int
    cache_hit_rate: float  # fraction of calls with any cache reads
    estimated_savings_usd: float
    avg_context_utilization_pct: float
    max_context_utilization_pct: float
    high_pressure_calls: int  # calls with context_utilization_pct >= 80
    trend: list[LensCacheTrendPoint]
    period_start: str
    period_end: str


# ---------------------------------------------------------------------------
# Session graph
# ---------------------------------------------------------------------------


class LensGraphNode(BaseModel):
    id: str
    event_type: str
    label: str
    status: str  # normal | warning | error | info
    metadata: dict[str, Any] | None = None


class LensGraphEdge(BaseModel):
    source: str
    target: str


class LensSessionGraphResponse(BaseModel):
    conversation_id: str
    nodes: list[LensGraphNode]
    edges: list[LensGraphEdge]
